### PR TITLE
Initial class to hold and parse Bucket metadata.

### PR DESCRIPTION
**Pull request by coryan**
This is the first version of the class, it is very incomplete,
but usable to write initial tests and prototypes.

### CODE REVIEW

File: storage/client/bucket_metadata.cc [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+         metadata_generation_ == rhs.metadata_generation_ and
+         name_ == rhs.name_ and project_number_ == rhs.project_number_ and
+         self_link_ == rhs.self_link_ and
+         storage_class_ == rhs.storage_class_ and
+         time_created_ == rhs.time_created_ and
+         time_updated_ == rhs.time_updated_;
+}
+
+std::ostream& operator<<(std::ostream& os, BucketMetadata const& rhs) {
+  // TODO() - convert back to JSON for a nicer format.
```
Comment (mbrukman) [inline storage/client/bucket_metadata.cc]: Should the TODO have a username or an issue number? Just a thought.

File: storage/client/bucket_metadata.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+/**
+ * Represents a Google Cloud Storage Bucket Metadata object.
+ *
+ * @warning This is an incomplete implementation to validate the design. It does
+ * not support changes to the metadata. It also lacks support for ACLs, CORS,
+ * encryption keys, lifecycle rules, and other features.
+ * TODO() - complete the implementation.
```
Comment (mbrukman) [inline storage/client/bucket_metadata.h]: Same question as above.

File: storage/client/bucket_metadata_test.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/client/bucket_metadata.h"
+#include <gmock/gmock.h>
+
+/// @test Verify that we parse JSON objects into BucketMetadata objects.
+TEST(BucketMetadataTest, Parse) {
+  std::string text = R"""({
+ "kind": "storage#bucket",
```
Comment (mbrukman) [inline storage/client/bucket_metadata_test.cc]: Would it make sense to indent these JSON literal lines 4 spaces for readability?

File: storage/client/bucket_metadata.cc [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/client/bucket_metadata.h"
+#include "storage/client/internal/nljson.h"
+#include "storage/client/internal/parse_rfc3339.h"
+
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+BucketMetadata BucketMetadata::ParseFromJson(std::string const& payload) {
```
Comment (mbrukman) [inline storage/client/bucket_metadata.cc]: What does this return if it cannot parse JSON, e.g., if the payload string is not valid JSON, or not all the fields are specified (is the metadata object valid if only a subset of the fields are set)?

File: storage/client/bucket_metadata.cc [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  auto json = storage::internal::nl::json::parse(payload);
+  BucketMetadata result{};
+  result.etag_ = json["etag"].get<std::string>();
+  result.id_ = json["id"].get<std::string>();
+  result.kind_ = json["kind"].get<std::string>();
```
Comment (mbrukman) [inline storage/client/bucket_metadata.cc]: What if `"kind"` does not exist in the `json` object?

File: storage/client/bucket_metadata_test.cc [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  EXPECT_EQ(1526758284L,
+            duration_cast<std::chrono::seconds>(
+                actual.time_updated().time_since_epoch())
+                .count());
+}
+
+/// @test Verify that we parse JSON objects into BucketMetadata objects.
+TEST(BucketMetadataTest, ParseWithLabels) {
+  std::string text = R"""({
+ "kind": "storage#bucket",
```
Comment (mbrukman) [inline storage/client/bucket_metadata_test.cc]: Same comment as above re: 4-space indent. Thoughts?

File: storage/client/bucket_metadata.cc (hunk 1 above)
Comment (coryan) [reply storage/client/bucket_metadata.cc]: Should be, fixed.

File: storage/client/bucket_metadata_test.cc (hunk 3 above)
Comment (coryan) [reply storage/client/bucket_metadata_test.cc]: Sure, done.

File: storage/client/bucket_metadata_test.cc (hunk 6 above)
Comment (coryan) [reply storage/client/bucket_metadata_test.cc]: Fixed.

**Comment by coryan** [comment]
PTAL.

File: storage/client/bucket_metadata.cc (hunk 4 above)
Comment (mbrukman) [reply storage/client/bucket_metadata.cc]: How would this work for a hypothetical customer who wants to disable exceptions?

File: storage/client/bucket_metadata.cc (hunk 5 above)
Comment (mbrukman) [reply storage/client/bucket_metadata.cc]: How would this work for a hypothetical customer who wants to disable exceptions?

File: storage/client/bucket_metadata.cc [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      std::stoll(json["projectNumber"].get_ref<std::string const&>());
+  result.self_link_ = json["selfLink"].get<std::string>();
+  result.storage_class_ = json["storageClass"];
+  result.time_created_ = storage::internal::ParseRfc3339(json["timeCreated"]);
+  result.time_updated_ = storage::internal::ParseRfc3339(json["updated"]);
+  return result;
+}
+
+bool BucketMetadata::operator==(BucketMetadata const& rhs) const {
+  return etag_ == rhs.etag_ and id_ == rhs.id_ and kind_ == rhs.kind_ and
```
[the remark this answers is not in the document]
Comment (houglum) [reply storage/client/bucket_metadata.cc]: On the etag_ thing... _sort of_.  

An etag isn't unique to a bucket.  In fact, I think the etag for a bucket is some hash/coding of the metageneration.  For two newly-created buckets, the etag for both is "CAE=".  After updating both, their etags are both "CAI=".  So, I'd throw etag up there next to metageneration in the comparison order :)

[the remark this answers is not in the document]
Comment (coryan) [reply storage/client/bucket_metadata.cc]: Thanks, I did not know how that worked obviously.  Fixed.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**